### ISSUE TITLE
Make foliage repositories buildable Nix

### ISSUE THREAD

**Issue by michaelpj** [andreabedini/foliage#46]
The main thing this would require is to be able to do all the fetching in a Nix-friendly way, which would require https://github.com/andreabedini/foliage/issues/45 and also some cleverness in working out what foliage needs, e.g. we might need something like:

```
$ foliage list-requirements
<url1> <hash1>
<url2> <hash2>
...
```

Then we could use IFD to make a derivation that prefetches everything into a directory, or use the intercepting-webserver trick.

**Comment by andreabedini** [comment]
:100:

I have been thinking about this. I think we could start by allowing optional hashes to meta.toml.

Questions/thoughts:
- What format for hashes? `sha256 = ...` ? Should we use the [SRI format](https://www.w3.org/TR/SRI/) like nix has started doing? Would that be more future proof?
- What is the hash of? The fetched tarball? (o repo à la nix? is there even a standard way?).
- Then we need to figure out how to tell foliage where the fetched inputs are. Maybe just with a decent convention? Switch the cache to be content addressed?
- Maybe we should integrate this into the metadata file and offer a way to produce it separately?

**Comment by michaelpj** [comment]
> We need to figure out how to tell foliage where the fetched inputs are.

Yeah, this is a bit annoying. I think it would be fine to have the answer be "pre-populate the cache" or something. It's some Nix code that we have to write, mostly. 

(Amusingly, this is actually a perfect case for recursive Nix (we replace `curl` with a script that calls `fetchurl` with the provided hash).)

> Maybe we should integrate this into the metadata file and offer a way to produce it separately?

Not sure what "this" is? The "prefetched source location"? I do think we might want something like that, although I think we need a top-level metadata file really for some repository-level options. That might be useful anyway for things like "default value of signing/writing metadata" etc.
